### PR TITLE
Add rate limiting to the retrofit module

### PR DESCRIPTION
**Pull request by cpilsworth**
Adds support for rate limiting http client calls as described in #97 

* Added RateLimiterCallAdapter
* Created internal DecoratedCall to minimise duplication
* Added tests for Retrofit Call without type parameters for better coverage
* Updated docs and included rate limiting examples

### CODE REVIEW

File: resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     * @param rateLimiter  {@link RateLimiter} to apply
+     * @param call            Call to decorate
+     * @param <T> Response type of call
+     * @return Original Call decorated with CircuitBreaker
+     */
+    static <T> Call<T> decorateCall(final RateLimiter rateLimiter, final Call<T> call) {
+        return new DecoratedCall<T>(call) {
+            @Override
+            public Response<T> execute() throws IOException {
+                Try.CheckedSupplier<Response<T>> restrictedSupplier = RateLimiter.decorateCheckedSupplier(rateLimiter, call::execute);
```
Comment (cpilsworth) [inline resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java]: Is this an idiomatic way of using Javaslang's Try @RobWin?

Comment (RobWin) [inline resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java]: Can you use Try.of(restrictedSupplier).recover(ex -> tooManyRequestsError()).get() ?

**Comment by RobWin** [comment]
Good work 👍

File: resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java (hunk 1 above)
Comment (cpilsworth) [inline resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java]: I have updated the code to this now:
```
public Response<T> execute() throws IOException {
                final Try.CheckedSupplier<Response<T>> restrictedSupplier = RateLimiter.decorateCheckedSupplier(rateLimiter, call::execute);
                return Try.of(restrictedSupplier)
                        .recover(ex -> Match(ex).of(
                                Case(instanceOf(RequestNotPermitted.class), tooManyRequestsError()),
                                Case(instanceOf(IllegalStateException.class), tooManyRequestsError()))
                        ).get();
            }
```
However, Try seems to want to always swallow the exceptions and return some Response<T>.  I really only want to consume the `RequestNotPermitted` and perhaps `IllegalStateException` exceptions so as not to change the interface for `Call` in decorating.  

The code above will give a MatchError on the a IOException from Call.execute - I would like to re-throw on the "else" match if possible.  Is there any way of doing this?

Comment (RobWin) [inline resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java]: Yes, Try might not be a good match for this type of use case. Might be simpler if you just use an imperative style.

**Comment by cpilsworth** [comment]
Thanks for reviewing @RobWin :eyeglasses:

**Pull request merged**